// _sula@1.0.0-beta.16@sula/es/render-plugin/index.js
import "antd/es/progress/style";
import _Progress from "antd/es/progress";
import "antd/es/badge/style";
import _Badge from "antd/es/badge";
import "antd/es/tag/style";
import _Tag from "antd/es/tag";
import "antd/es/input/style";
import _Input from "antd/es/input";
import Card from './card';
import Text from './text'; // import Icon from './icon';

import Icon, { IconPlugin } from './icon';
import Button, { ButtonPlugin, LinkPlugin } from './button'; // only plugin

import RowSelection from './rowSelection';
import { registerRenderPlugin } from './plugin';

function registerRenderPlugins() {
  registerRenderPlugin('card')(Card);
  registerRenderPlugin('div')('div');
  registerRenderPlugin('inputgroup')(_Input.Group);
  registerRenderPlugin('text')(Text);
  registerRenderPlugin('button', ['autoLoading'])(ButtonPlugin, true);
  registerRenderPlugin('link', ['autoLoading'])(LinkPlugin, true);
  registerRenderPlugin('icon', ['autoLoading'])(IconPlugin, true);
  registerRenderPlugin('rowSelection')(RowSelection, true);
  registerRenderPlugin('tag')(_Tag);
  registerRenderPlugin('badge')(_Badge);
  registerRenderPlugin('progress')(_Progress);
}

export { Card, Text, Icon, Button, registerRenderPlugins, registerRenderPlugin };
