// _sula@1.0.0-beta.16@sula/es/_util/action-logger.js
var actionColors = {
  action: '#9e9e9e',
  beforePass: '#5b8c00',
  beforeStop: '#eb2f96',
  actionError: '#f5222d',
  actionSuccess: '#389e0d'
};
var actionLoggers = {
  action: function action(type, payload) {
    console.log("%caction %c".concat(payload), getStyle(actionColors[type]), getStyle('#1890ff'));
  },
  beforePass: function beforePass(type) {
    console.log("    %cbefore pass", getStyle(actionColors[type]));
  },
  beforeStop: function beforeStop(type) {
    console.log("    %cbefore stop", getStyle(actionColors[type]));
  },
  actionError: function actionError(type, payload) {
    console.log("    %cerror", getStyle(actionColors[type]), payload);
  },
  actionSuccess: function actionSuccess(type, payload) {
    console.log("    %csuccess", getStyle(actionColors[type]), payload);
  }
};
export default function actionLogger(type, payload) {
  // @ts-ignore
  if (process.env.SULA_LOGGER === "action" || process.env.SULA_LOGGER === "all") {
    actionLoggers[type](type, payload);
  }
}

function getStyle(color) {
  var isBold = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var style = "color: ".concat(color, ";");

  if (isBold) {
    style += "font-weight:bold;";
  }

  return style;
}
