// _sula@1.0.0-beta.16@sula/es/field-plugin/index.js
import "antd/es/slider/style";
import _Slider from "antd/es/slider";
import "antd/es/radio/style";
import _Radio from "antd/es/radio";
import "antd/es/checkbox/style";
import _Checkbox from "antd/es/checkbox";
import "antd/es/switch/style";
import _Switch from "antd/es/switch";
import "antd/es/rate/style";
import _Rate from "antd/es/rate";
import "antd/es/input-number/style";
import _InputNumber from "antd/es/input-number";
import "antd/es/input/style";
import _Input from "antd/es/input";
import Select from './select';
import Upload from './upload';
import Cascader from './cascader';
import CheckboxGroup from './checkboxgroup';
import RadioGroup from './radiogroup';
import DatePicker from './datepicker';
import RangePicker from './rangepicker';
import TimePicker from './timepicker';
import { registerFieldPlugin } from './plugin';

function registerFieldPlugins() {
  registerFieldPlugin('input')(_Input);
  registerFieldPlugin('textarea')(_Input.TextArea);
  registerFieldPlugin('password')(_Input.Password);
  registerFieldPlugin('inputnumber')(_InputNumber);
  registerFieldPlugin('rate')(_Rate);
  registerFieldPlugin('switch')(_Switch);
  registerFieldPlugin('checkbox')(_Checkbox);
  registerFieldPlugin('radio')(_Radio);
  registerFieldPlugin('slider')(_Slider);
  registerFieldPlugin('cascader')(Cascader, true);
  registerFieldPlugin('datepicker')(DatePicker);
  registerFieldPlugin('rangepicker')(RangePicker);
  registerFieldPlugin('timepicker')(TimePicker);
  registerFieldPlugin('upload')(Upload, false, true);
  registerFieldPlugin('select')(Select, true);
  registerFieldPlugin('checkboxgroup')(CheckboxGroup, true);
  registerFieldPlugin('radiogroup')(RadioGroup, true);
}

export { Select, CheckboxGroup, TimePicker, DatePicker, RangePicker, Cascader, RadioGroup, Upload, registerFieldPlugins, registerFieldPlugin };
